Tests/ApolloTests/FetchQueryTests.swift

```diff
@@ -364,4 +364,95 @@ class FetchQueryTests: XCTestCase, CacheDependentTesting {
     expect(cacheMissResult).to(beNil())
   }
 
+  // MARK: Concurrency Load Testing
+
+  func test_concurrentFetchesWritingToStore_avoidsDeadlockAndDataRaces() async throws {
+    class HeroNameSelectionSet: MockSelectionSet, @unchecked Sendable {
+      override class var __selections: [Selection] {
+        [
+          .field("hero", Hero.self)
+        ]
+      }
+
+      class Hero: MockSelectionSet, @unchecked Sendable {
+        override class var __selections: [Selection] {
+          [
+            .field("__typename", String.self),
+            .field("name", String.self),
+          ]
+        }
+      }
+    }
+
+    let serverRequestExpectation =
+      await server.expect(MockQuery<HeroNameSelectionSet>.self) { request in
+        [
+          "data": [
+            "hero": [
+              "name": "Luke Skywalker",
+              "__typename": "Human",
+            ]
+          ]
+        ]
+      }
+    serverRequestExpectation.assertForOverFulfill = false
+
+    // When: Execute multiple queries concurrently from different threads
+    // Using networkOnly policy - Apollo will still write responses to cache,
+    // triggering concurrent cache merge operations that cause the crash
+    let numberOfConcurrentQueries = 1000 // Increase to make crash more likely
+
+    // Use actor-safe counter to track completed tasks
+    actor TaskCounter {
+      private var completedCount = 0
+
+      func increment() {
+        completedCount += 1
+      }
+
+      func getCount() -> Int {
+        completedCount
+      }
+    }
+
+    let counter = TaskCounter()
+
+    // Execute all queries concurrently and wait for all to complete
+    await withTaskGroup(of: Void.self) { group in
+      // Add all tasks to the group
+      for _ in 0..<numberOfConcurrentQueries {
+        group.addTask { [client] in
+          do {
+            // Use networkOnly - Apollo will still write to cache, triggering merge operations
+            // Execute on different threads to maximize concurrency
+            _ = try await client!.fetch(
```

Comment:
At the moment this test suite just uses a standard `ApolloClient` and therefore a default interceptor list. If that was to ever change and the caching interceptor was removed would this test still push the boundaries of concurrency we're expecting? If not, is there a sanity validation check we should add somewhere in this test to ensure the caching interceptor is in the request chain?

Reply:
With 2.0 there is no way to remove the caching interceptor. The `InterceptorProvider` requires on. You would need to replace the default cache interceptor with one that did nothing in order to break this test.